File: ben/haskell/Q11-20-lists-cont/p19.hs
-- With prelude
rotate :: [a] -> Int -> [a]
rotate list nb
    | nb < 0 = rotate list (length list + nb)
    | otherwise = drop nb list ++ take nb list

-- With list recursion
rotate' :: [a] -> Int -> [a]
rotate' list@(x:xs) nb
    | nb == 0 = list
    | nb < 0 = rotate (last list : init list) (nb + 1)
    | otherwise = rotate (xs ++ [x]) (nb -1)

File: ben/prolog/P01-P28-lists/p19.pl
% Rotate a list N places to the left.

rotate(L, 0, L).
rotate([X|Xs], N0, R) :-
    N0 > 0, 
    N1 is N0 - 1,
    append(Xs, [X], L),
    rotate(L, N1, R).
rotate(L, N0, R) :-
    N0 < 0, 
    N1 is N0 + 1,
    last(L, X),
    init(L, Xs), 
    rotate([X|Xs], N1, R).
    
init([_], []).
init([X|Xs], [X|L]) :-
    init(Xs, L).
